File: mobile/swift/Smoke.swift
import Foundation

@main
struct DLPSmoke {
    static func main() throws {
        let runtime = try DLPRuntime(eq: 1, neq: 2, top: 3, seed: 1)
        defer { runtime.close() }
        try runtime.addTerm(1, orderKey: "urn:seed")
        try runtime.addTerm(2, orderKey: "urn:a:🛣️")
        try runtime.addFact(10, arguments: [2])
        try runtime.addUnaryRule(head: 11, body: 10)
        let stats = try runtime.materialize()
        guard stats.complete, try runtime.facts().contains(where: { $0.predicate == 11 && $0.arguments == [2] }) else {
            throw DLPError(description: "Missing inferred fact")
        }
        let distance = try DLPRuntime.wgs84Distance(longitude1: 0, latitude1: 0, longitude2: 1, latitude2: 0)
        guard abs(distance - 111319.49079327357) < 1e-6 else { throw DLPError(description: "Wrong WGS84 result") }
        runtime.close()
        do {
            _ = try runtime.materialize()
            throw DLPError(description: "Closed runtime accepted an operation")
        } catch let error as DLPError where error.description == "DLP runtime is closed" {}
        print("PASS Swift: native inference, Unicode term, bounded fact page, WGS84 distance, closed-handle rejection")
        let windows = try DLPWindowStore(width: 10, allowedLateness: 20, context: Data("map-v1".utf8))
        defer { windows.close() }
        _ = try windows.upsert(DLPWindowEvent(id: Data("a".utf8), key: Data("car".utf8), payload: Data([0]), time: 0))
        _ = try windows.upsert(DLPWindowEvent(id: Data("b".utf8), key: Data("car".utf8), payload: Data([1]), time: 5))
        let enteredWindow = try windows.advance(to: 10)
        guard enteredWindow.added.count == 2 else { throw DLPError(description: "Window publication differs") }
        let resumed = try DLPWindowStore.restore(windows.checkpoint(), expectedContext: Data("map-v1".utf8))
        defer { resumed.close() }
        let expired = try resumed.advance(to: 100)
        guard expired.removed.count == 2, expired.historyRemoved.count == 1 else {
            throw DLPError(description: "Window expiry/predecessor differs")
        }
        print("PASS Swift native windows: identified events, publication, checkpoint/resume, idle expiry, predecessor retention")
    }
}

File: mobile/android/Smoke.kt
package org.example.dlp

import kotlin.math.abs

fun main() {
    val runtime = NativeRuntime(1, 2, 3, 1)
    runtime.use {
        it.addTerm(1, "urn:seed")
        it.addTerm(2, "urn:a:🛣️")
        it.addFact(10, longArrayOf(2))
        it.addUnaryRule(11, 10)
        check(it.materialize().complete)
        check(it.facts().any { row -> row.predicate == 11L && row.arguments == listOf(2L) })
        check(abs(NativeRuntime.wgs84Distance(0.0, 0.0, 1.0, 0.0) - 111319.49079327357) < 1e-6)
        check(runCatching { it.facts(limit = 4097) }.exceptionOrNull() is IllegalArgumentException)
    }
    check(runCatching { runtime.materialize() }.exceptionOrNull() is IllegalStateException)
    runtime.close()
    println("PASS Kotlin/JNI: inference, Unicode term, bounded fact page, WGS84 distance, closed-handle rejection")
    NativeWindowStore(10, allowedLateness = 20, context = "map-v1".toByteArray()).use { window ->
        window.upsert(NativeWindowStore.Event("a".toByteArray(), "car".toByteArray(), byteArrayOf(0), 0))
        window.upsert(NativeWindowStore.Event("b".toByteArray(), "car".toByteArray(), byteArrayOf(1), 5))
        check(window.advance(10).added == 2L)
        NativeWindowStore.restore(window.checkpoint(), "map-v1".toByteArray()).use { restored ->
            val change = restored.advance(100)
            check(change.removed == 2L && change.historyRemoved == 1L)
        }
    }
    println("PASS Kotlin/JNI native windows: events, change counts, checkpoint/resume, idle expiry, predecessor retention")
}
